core/cache/charmconfigwatcher.go

```diff
@@ -0,0 +1,279 @@
+		_ = w.Stop()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// init determines baseline master configuration, branch tracking settings
+// and the configuration hash for the watcher's unit.
+// It then closes the init channel to indicate the watcher is operational.
+func (w *CharmConfigWatcher) init(model charmConfigModel) error {
+	app, err := model.Application(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.masterSettings = app.Config()
+
+	for _, b := range model.Branches() {
+		unitNames := b.AssignedUnits()[w.appName]
+		if len(unitNames) == 0 {
+			continue
+		}
+
+		// A unit can only track one branch at a time,
+		// so if we find one, we are done.
+		if set.NewStrings(unitNames...).Contains(w.unitName) {
+			w.branchName = b.Name()
+			w.branchDeltas = b.Config()[w.appName]
+			break
+		}
+	}
+
+	if _, err := w.setConfigHash(); err != nil {
+		return errors.Trace(err)
+	}
+
+	close(w.initComplete)
+	return nil
+}
+
+// appConfigChanged is called when a message is received indicating changed
+// application master charm configuration.
+func (w *CharmConfigWatcher) appConfigChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	hashCache, ok := msg.(*hashCache)
+	if !ok {
+		logger.Errorf("programming error; application config message was not of expected type, *hashCache")
+		return
+	}
+
+	w.masterSettings = hashCache.config
+	w.checkConfig()
+}
+
+// branchChanged is called when we receive a message to say that a branch has
+// been updated in the cache.
+// If it is the branch that this watcher's unit is tracking,
+// check if the latest config delta warrants a notification.
+func (w *CharmConfigWatcher) branchChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	b, okUnit := msg.(Branch)
+	if !okUnit {
+		logger.Errorf("programming error; branch change message was not of expected type, Branch")
+		return
+	}
+
+	if w.branchName != "" {
+		// If we are tracking a different branch to this one, ignore it.
+		if w.branchName != b.Name() {
+			return
+		}
+	} else {
+		// We do not know if we are tracking this branch, so find out.
+		units := b.AssignedUnits()[w.appName]
+		if len(units) == 0 {
+			return
+		}
+		if !set.NewStrings(units...).Contains(w.unitName) {
+			return
+		}
+		w.branchName = b.Name()
+	}
+
+	w.branchDeltas = b.Config()[w.appName]
+	w.checkConfig()
+}
+
+// branchDeleted is called when we receive a message to say that a branch has
+// been removed from the cache.
+// If this watcher's unit was tracking the branch, clean the branch-based
+// details and check if the resulting settings warrant a notification.
+func (w *CharmConfigWatcher) branchDeleted(topic string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	name, okUnit := msg.(string)
+	if !okUnit {
+		logger.Errorf("programming error; branch deleted message was not of expected type, string")
+		return
+	}
+
+	if w.branchName != name {
```

Comment:
Don't you need to check if `w.branchName` is empty like you do with the branch-changed logic above?

Reply:
Ooh, nice one. Should also flex that in a test.

Reply:
Actually no. Because if the watcher is tracking no branch (empty) or a different one to the deleted one, we don't care.